accepted/00000-maintenance-windows.md

```diff
@@ -0,0 +1,99 @@
+- Feature Name: Maintenance Windows
+- Start Date: 2020-03-31
+- RFC PR: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Define Maintenance Windows in Uyuni and assign them to systems
+
+# Motivation
+[motivation]: #motivation
+
+Customers have company policies to perform changes on systems only at specific points in time.
+This is typically called "Maintenance Window".
+
+Uyuni should be able to define them and prevent executing of actions outside of such a window.
+
+# Architecture
+[architecture]: #architecture
+
+A system is in "Maintenance Mode" or not.
+When a system is in "Maintenance Mode" no action can be executed.
+Setting the Maintenance Mode "on" and "off" requires Organization Admin permissions.
+A "Maintenance Window" does nothing else than turning Maintenance Mode "on" when the window starts
+and "off" when it ends.
+
+Defining Maintenance Windows requires Organization Admin permissions.
+
+Maintenance Windows are defined on system level.
+Every System can get 1 Schedule assigned.
+A Schedule can be used for multiple systems.
+Inside of the Schedule, multiple Maintenance Windows can be defined.
+They can be defined as individual or as recurring events.
+
+To assign a Schedule to multiple systems, SSM should be used.
+
+
+# Detailed design
+[design]: #detailed-design
+
+## Implement a Maintenance Mode
+
+Check, how System Locking is implemented. Copy the mechanism for Maintenance Mode.
+Does it make sense to use System Locking as "Maintenance Mode"?
```

Comment:
Publishing a comment about my research of system locking. There are 3 types of locking in Uyuni:

### Package locking
(in the web UI: `Software -> Packages -> Lock`)
- this allows locking some package versions on the system (probably uses zypper locking mechanism under the hood?)
- looks like we need something more general

### System locking (traditional systems only)

(in the Web UI, go to system details, click `lock system`)

- systems locked in such way will not allow running _some_ actions:
- the (black)list of forbidden actions is set in the DB: `select * from rhnActionType where unlocked_only = 'Y';`
- the blacklist is not customizable by the user
- all actions (even the blacklisted) can still be scheduled, when the system is locked
- the actions not in the blacklist are run normally
- as for the blacklisted actions: as soon as the system is unlocked, these scheduled actions are run (i.e. on next execution of the `rhn_check`, supposing the system is unlocked at the time of execution)
- implementation: system lock is an object in the database, there is a `1:1` relation with the system (or server) object
- **problems**
  - if we use it for the maintenance mode, we could collide with the existing locks (the admin can set the lock, independent of any maintenance window. Maybe we don't want to unlock the system that was locked explicitly by some administrator). This could be solved by documenting the behavior ("system locks can be overridden by the maintenance windows mechanism").

### Blackout mode (minions only)
- i.e. locking on "salt level"
- Implementation via a formula (see `system-lock-formula` RPM)
- when the formula is assigned to a system (group, org), the salt module calls will not be executed on the system(s) 
- difference to the system lock for trad. systems: when the formula is assigned, any actions run will **fail**
- allows whitelisting of allowed actions (`test.ping`, `grains.items`, `grains.item`)
- **problems**:
  - same as the previous case

### Q&A
- Q: we need to decide, how we should react when an action is executed outside of the maintenance window
  - A1: Postpone
  - A2: Fail
  - ~A3~: Don't allow executing that action at all. That would be very hard to achieve: actions can be scheduled for future. Maintenance window that forbids this action can be created **after** the action is scheduled.

- Q:  do we want to support both traditional clients and salt minions? if yes, we need to unify the fail/postpone semantics above
  - A: We focus primarily on minions. We could use the blackout mechanism for minions and go with the "fail" semantics above.

- Q: Can we actually use the blackout mechanism for minions?
  - pros:
    - already implemented mechanism
    - works on the salt modules level, which should be powerful and flexible enough
  - cons:
    - can collide with existing code that uses blacklisting (we would need a special handling for CaaSP nodes, which should always stay in blackout mode).
    - implementation via formula: it inherits the common formula integration drawbacks, which we might take into account (assignment on filesystems, not in the database: slower querying, no support for migration scripts)

Reply:
as we decided to not enforce the maintenance mode on the system itself in the first iteration I will mark this discussion as resolved.
This will come back when we discuss the 2nd iteration.